File: COMPILADOR/kokoslan_/src/src/java/ast/KoKoOperation.java
/**
 @author loriacarlos@gmail.com
 
*/
package src.java.ast;
import java.util.*;
import java.io.*;
import src.java.eval.KoKoContext;
import src.java.eval.KoKoValue;
import src.java.exception.KoKoEvalException;


public class KoKoOperation implements KoKoAst{
   protected List<KoKoAst> operands;
   protected KoKoAst operator;
   
   public KoKoOperation(KoKoAst operator, List<KoKoAst> operands){
	  this.operator = operator;
      this.operands = operands;
	  
   }
   public KoKoOperation(KoKoAst operator){
	  this(operator, new ArrayList<KoKoAst>	());
	  
   }
   public void addOperand(KoKoAst x){
	   this.operands.add(x);
   }
   public void genCode(PrintStream out){
	   this.operands.get(0).genCode(out);
       this.operands.stream()
	                .skip(1)
	                .forEach( t -> {
						  this.operator.genCode(out);
					      t.genCode(out);
					   });
   }
   public KoKoValue eval(KoKoContext ctx){
	   throw new KoKoEvalException("KoKoOperation: eval not implemented");
	   
   }
   
}

File: COMPILADOR/kokoslan/src/KoKoKotlin/ast/KoKoOperation.kt
/**
 * @author loriacarlos@gmail.com
 */
package KoKoKotlin.ast

import KoKoKotlin.eval.KoKoContext
import KoKoKotlin.eval.KoKoValue
import KoKoKotlin.exception.KoKoEvalException
import java.util.*
import java.io.*

import src.java.ast.KoKoAst



open class KoKoOperation @JvmOverloads constructor(protected var operator: KoKoAst, protected var operands: MutableList<KoKoAst> = ArrayList()) : KoKoAst {
    fun addOperand(x: KoKoAst) {
        this.operands.add(x)
    }

    override fun genCode(out: PrintStream) {
        this.operands[0].genCode(out)
        this.operands.stream()
                .skip(1)
                .forEach { t ->
                    this.operator.genCode(out)
                    t.genCode(out)
                }
    }

    override fun eval(ctx: KoKoContext): KoKoValue {
        throw KoKoEvalException("KoKoOperation: eval not implemented")

    }

}